File: pdui_flutter/lib/widgets/pdui_scaffold.dart
import 'package:flutter/material.dart';

import '../core/pdui_expression_resolver.dart';
import '../proto_out/pdui-proto-out.pb.dart';
import 'pdui_app_bar.dart';
import 'pdui_widget.dart';

class PduiScaffold extends PduiWidget {
  PBExpression pbScaffold;
  PduiScaffold(this.pbScaffold);

  @override
  Scaffold resolve() {
    return Scaffold(
      key: pbScaffold.key != "" ? Key(pbScaffold.key) : null,
      appBar: PduiAppBar(pbScaffold.scaffold.appBar).resolve(),
      body: PduiExpressionResolver.parse(pbScaffold.scaffold.body),
    );
  }
}

File: pdui-express/lib/widgets/PDUIScaffold.ts
import { PDUIWidget } from "../models/PDUIWidget";
import { PBScaffold, type PBExpression } from "./proto-out/widgets";

export class PDUIScaffold extends PDUIWidget {
    appBar: PDUIWidget;
    body: PDUIWidget;

    constructor({
        key,
        appBar,
        body,
    }: { key?: string; appBar: PDUIWidget; body: PDUIWidget }) {
        super({
            key,
            identity: "scaffold",
            payload: {
                oneofKind: "scaffold",
                scaffold: { appBar: appBar.toPB(), body: body.toPB() },
            },
        });

        this.appBar = appBar;
        this.body = body;
    }

    override toPB(): PBScaffold {
        return {
            appBar: this.appBar.toPB(),
            body: this.body.toPB(),
        };
    }

    override toExpression(): PBExpression {
        return {
            key: this.key,
            identity: this.identity,
            payload: {
                oneofKind: "scaffold",
                scaffold: {
                    appBar: this.appBar.toExpression(),
                    body: this.body.toExpression(),
                },
            },
        };
    }
}
